File: code/old/trilateration/trilateration.js
function determinant3(mat) {
    const det = mat[0][0] * (mat[1][1] * mat[2][2] - mat[1][2] * mat[2][1]) -
                mat[0][1] * (mat[1][0] * mat[2][2] - mat[1][2] * mat[2][0]) +
                mat[0][2] * (mat[1][0] * mat[2][1] - mat[1][1] * mat[2][0]);
    console.log("Le déterminant est " + det);
    return det;
}

function inverse(mat) {
    const inv = [[], [], []];
    const det = determinant3(mat);

    if (det === 0) {
        console.log("Problème chef");
        return inv;
    }

    const fact = 1.0 / det;

    inv[0][0] = (mat[1][1] * mat[2][2] - mat[1][2] * mat[2][1]) * fact;
    inv[0][1] = (mat[0][2] * mat[2][1] - mat[0][1] * mat[2][2]) * fact;
    inv[0][2] = (mat[0][1] * mat[1][2] - mat[0][2] * mat[1][1]) * fact;
    inv[1][0] = (mat[1][2] * mat[2][0] - mat[1][0] * mat[2][2]) * fact;
    inv[1][1] = (mat[0][0] * mat[2][2] - mat[0][2] * mat[2][0]) * fact;
    inv[1][2] = (mat[0][2] * mat[1][0] - mat[0][0] * mat[1][2]) * fact;
    inv[2][0] = (mat[1][0] * mat[2][1] - mat[1][1] * mat[2][0]) * fact;
    inv[2][1] = (mat[0][1] * mat[2][0] - mat[0][0] * mat[2][1]) * fact;
    inv[2][2] = (mat[0][0] * mat[1][1] - mat[0][1] * mat[1][0]) * fact;

    return inv;
}

function printMatrix(mat) {
    for (let i = 0; i < 3; i++) {
        console.log(mat[i].join(' '));
    }
}

function product(mat1, mat2) {
    const result = [[], [], []];

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            result[i][j] = 0;
            for (let k = 0; k < 3; k++) {
                result[i][j] += mat1[i][k] * mat2[k][j];
            }
        }
    }

    return result;
}

function transpose(mat) {
    const transpose = [[], [], []];
    transpose[0][0] = mat[0][0]; transpose[0][1] = mat[1][0]; transpose[2][0] = mat[0][2];
    transpose[1][1] = mat[1][1]; transpose[0][2] = mat[2][0]; transpose[1][0] = mat[0][1];
    transpose[2][2] = mat[2][2]; transpose[1][2] = mat[2][1]; transpose[2][1] = mat[1][2];

    return transpose;
}

function aloisisaffiching(vector) {
    for (let i = 0; i < 3; i++) {
        console.log(vector[i]);
    }
}

function merciAloisFranchementBouh(mat, vector) {
    const result = [0, 0, 0];

    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            result[i] += mat[i][j] * vector[j];
        }
    }

    return result;
}

function main() {
    const da = 2.0;
    const db = 3.0;
    const dc = 4.0;
    const dd = 75.0;

    const xa = 1.0;
    const ya = 21.0;
    const za = 3.0;

    const oa = dist(xa, ya, za);

    const xb = 1.0;
    const yb = 23.0;
    const zb = 3.0;

    const ob = dist(xb, yb, zb);

    const xc = 17.0;
    const yc = 29.0;
    const zc = 3.0;

    const oc = dist(xc, yc, zc);

    const xd = 2.0;
    const yd = 1.0;
    const zd = 30.0;

    const od = dist(xd, yd, zd);

    const u = [
        (da * da - db * db + ob - oa) * 0.5,
        (da * da - dc * dc + oc - oa) * 0.5,
        (da * da - dd * dd + od - oa) * 0.5
    ];

    aloisisaffiching(u);

    const p = [
        [xb - xa, xc - xa, xd - xa],
        [yb - ya, yc - ya, yd - ya],
        [zb - za, zc - za, zd - za]
    ];

    printMatrix(p);

    const pete = transpose(inverse(p));

    printMatrix(pete);

    const final = merciAloisFranchementBouh(pete, u);
    aloisisaffiching(final);
}

function dist(x, y, z) {
    return x * x + y * y + z * z;
}

main();

File: code/old/trilateration/trilateration.c
#include <stdarg.h>
#include <stdio.h>

typedef struct {
    float data[3][3];
} matrix;

typedef struct {
    float data[3];
} aloisismean;



float determinant3(matrix mat){

    float det =  mat.data[0][0] * (mat.data[1][1] * mat.data[2][2] - mat.data[1][2] * mat.data[2][1]) -
           mat.data[0][1] * (mat.data[1][0] * mat.data[2][2] - mat.data[1][2] * mat.data[2][0]) +
           mat.data[0][2] * (mat.data[1][0] * mat.data[2][1] - mat.data[1][1] * mat.data[2][0]);
    printf("Le déterminant est %f \n",det);
    return det;
    };

matrix inv(matrix mat){
    matrix inv; 
    float det = determinant3(mat);

    if (det==0){
        printf("Problème chef");
        return inv;
    }

    float fact = (1.0)/det;

    inv.data[0][0] = (mat.data[1][1] * mat.data[2][2] - mat.data[1][2]*mat.data[2][1])*fact;
    inv.data[0][1] = (mat.data[0][2] * mat.data[2][1] - mat.data[0][1] *mat.data[2][2]) * fact;
    inv.data[0][2] = (mat.data[0][1] * mat.data[1][2]- mat.data[0][2] * mat.data[1][1]) * fact;
    inv.data[1][0] = (mat.data[1][2] *mat.data[2][0] - mat.data[1][0] * mat.data[2][2]) * fact;
    inv.data[1][1] = (mat.data[0][0] * mat.data[2][2] - mat.data[0][2] * mat.data[2][0]) * fact;
    inv.data[1][2] = (mat.data[0][2] * mat.data[1][0] -mat.data[0][0]*mat.data[1][2]) * fact;
    inv.data[2][0] = (mat.data[1][0] * mat.data[2][1] - mat.data[1][1]* mat.data[2][0]) * fact;
    inv.data[2][1] = (mat.data[0][1] * mat.data[2][0] - mat.data[0][0] * mat.data[2][1]) * fact;
    inv.data[2][2] = (mat.data[0][0] * mat.data[1][1] - mat.data[0][1] * mat.data[1][0]) * fact;

    return inv;
    

};

void affiche(matrix mat) {
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            printf("%.2f ", mat.data[i][j]);
        }
        printf("\n");
    }
};

matrix produit(matrix mat1, matrix mat2) {
    matrix result;

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.data[i][j] = 0;
            for (int k = 0; k < 3; k++) {
                result.data[i][j] += mat1.data[i][k] * mat2.data[k][j];
            }
        }
    }

    return result;
};

matrix transpose(matrix mat){
    matrix transpose;
    transpose.data[0][0]=mat.data[0][0];transpose.data[0][1]=mat.data[1][0];transpose.data[2][0]=mat.data[0][2];
    transpose.data[1][1]=mat.data[1][1];transpose.data[0][2]=mat.data[2][0];transpose.data[1][0]=mat.data[0][1];
    transpose.data[2][2]=mat.data[2][2];transpose.data[1][2]=mat.data[2][1];transpose.data[2][1]=mat.data[1][2];

    return transpose;
};

float dist(float x, float y, float z){
    return x*x+y*y+z*z;
};

void aloisisaffiching(aloisismean vector) {
    for (int i = 0; i < 3; i++) {
        printf("%f\n", vector.data[i]);
    }
};

aloisismean merciAloisFranchementBouh(matrix mat, aloisismean vector) {
    aloisismean result = {{0, 0, 0}};

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            result.data[i] += mat.data[i][j] * vector.data[j];
        }
    }

    return result;
}



int main(char* argv){

    //Test
    /*matrix matrice;

    matrice.data[0][0] = 1.0; matrice.data[0][1] = 2.0; matrice.data[0][2] = 3.0;
    matrice.data[1][0] = 4.0; matrice.data[1][1] = 2.0; matrice.data[1][2] = 6.0;
    matrice.data[2][0] = 7.0; matrice.data[2][1] = 8.0; matrice.data[2][2] = 9.0;

    determinant3(matrice);
    matrix inverse = inv(matrice);
    affiche(inverse);
    matrix id = produit(matrice,inverse);
    affiche(id);*/

    //Fin Test

    /*
    float da = argv[2];
    float db = argv[3];
    float dc = argv[4];
    float dd = argv[5];
    */

    float da = 2.0;
    float db = 3.0;
    float dc = 4.0;
    float dd = 75.0;
    printf("%f",da);

    float xa = 1.0;
    float ya = 21.0;
    float za = 3.0; 

    float oa = dist(xa,ya,za);

    float xb = 1.0;
    float yb = 23.0;
    float zb = 3.0; 

    float ob = dist(xb,yb,zb);

    float xc = 17.0;
    float yc = 29.0;
    float zc = 3.0; 

    float oc = dist(xc,yc,zc);

    float xd = 2.0;
    float yd = 1.0;
    float zd = 30.0; 

    float od = dist(xd,yd,zd);


    aloisismean u;
    u.data[0]=(da*da-db*db+ob-oa)*0.5;
    u.data[1]=(da*da-dc*dc+oc-oa)*0.5;
    u.data[2]=(da*da-dd*dd+od-oa)*0.5;

    aloisisaffiching(u);

    matrix p;

    p.data[0][0] = xb-xa; p.data[0][1] = xc-xa; p.data[0][2] = xd-xa;
    p.data[1][0] = yb-ya; p.data[1][1] = yc-ya; p.data[1][2] = yd-ya;
    p.data[2][0] = zb-za; p.data[2][1] = zc-za; p.data[2][2] = zd-za;

    affiche(p);

    matrix pete = transpose(inv(p));

    affiche(pete);

    aloisismean final = merciAloisFranchementBouh(pete, u);
    aloisisaffiching(final);

    return 0; 
};